File: workflow-sdk-go/client/grpc.go
package client

import (
	"context"
	"encoding/json"
	"fmt"

	workflowv1 "github.com/batnam/rochallor-engine/workflow-sdk-go/internal/gen/workflow/v1"
	"google.golang.org/grpc"
	"google.golang.org/grpc/credentials/insecure"
	"google.golang.org/protobuf/encoding/protojson"
	"google.golang.org/protobuf/types/known/structpb"
)

// GrpcClient implements EngineClient via gRPC.
// Proto stubs are vendored into internal/gen/workflow/v1 to avoid a circular
// dependency with the engine module. Run make proto-sync in the engine to
// regenerate and copy updated stubs.
type GrpcClient struct {
	conn     *grpc.ClientConn
	workerID string
}

// NewGrpc dials target and returns a GrpcClient.
func NewGrpc(target, workerID string, opts ...grpc.DialOption) (*GrpcClient, error) {
	if len(opts) == 0 {
		opts = []grpc.DialOption{grpc.WithTransportCredentials(insecure.NewCredentials())}
	}
	conn, err := grpc.NewClient(target, opts...)
	if err != nil {
		return nil, fmt.Errorf("grpc dial %q: %w", target, err)
	}
	return &GrpcClient{conn: conn, workerID: workerID}, nil
}

// Close closes the underlying gRPC connection.
func (c *GrpcClient) Close() error { return c.conn.Close() }

// PollJobs implements EngineClient via the WorkflowEngine.PollJobs RPC.
func (c *GrpcClient) PollJobs(ctx context.Context, req PollJobsRequest) ([]Job, error) {
	stub := workflowv1.NewWorkflowEngineClient(c.conn)
	resp, err := stub.PollJobs(ctx, &workflowv1.PollJobsRequest{
		WorkerId: req.WorkerID,
		JobTypes: req.JobTypes,
		MaxJobs:  int32(req.MaxJobs),
	})
	if err != nil {
		return nil, fmt.Errorf("grpc PollJobs: %w", err)
	}
	return protoJobsToSDK(resp.Jobs), nil
}

// CompleteJob implements EngineClient via the WorkflowEngine.CompleteJob RPC.
func (c *GrpcClient) CompleteJob(ctx context.Context, jobID, workerID string, variablesToSet map[string]any) error {
	var vars *structpb.Struct
	if len(variablesToSet) > 0 {
		var err error
		vars, err = marshalStruct(variablesToSet)
		if err != nil {
			return fmt.Errorf("grpc CompleteJob: marshal variables: %w", err)
		}
	}
	stub := workflowv1.NewWorkflowEngineClient(c.conn)
	_, err := stub.CompleteJob(ctx, &workflowv1.CompleteJobRequest{
		JobId:          jobID,
		WorkerId:       workerID,
		VariablesToSet: vars,
	})
	if err != nil {
		return fmt.Errorf("grpc CompleteJob: %w", err)
	}
	return nil
}

// FailJob implements EngineClient via the WorkflowEngine.FailJob RPC.
func (c *GrpcClient) FailJob(ctx context.Context, jobID, workerID, errorMessage string, retryable bool) error {
	stub := workflowv1.NewWorkflowEngineClient(c.conn)
	_, err := stub.FailJob(ctx, &workflowv1.FailJobRequest{
		JobId:        jobID,
		WorkerId:     workerID,
		ErrorMessage: errorMessage,
		Retryable:    retryable,
	})
	if err != nil {
		return fmt.Errorf("grpc FailJob: %w", err)
	}
	return nil
}

// protoJobsToSDK maps a slice of proto Job messages to SDK Job structs.
func protoJobsToSDK(jobs []*workflowv1.Job) []Job {
	out := make([]Job, 0, len(jobs))
	for _, j := range jobs {
		out = append(out, protoJobToSDK(j))
	}
	return out
}

// protoJobToSDK maps a single proto Job to an SDK Job.
func protoJobToSDK(j *workflowv1.Job) Job {
	sdkJob := Job{
		ID:               j.Id,
		JobType:          j.JobType,
		InstanceID:       j.InstanceId,
		StepExecutionID:  j.StepExecutionId,
		RetriesRemaining: int(j.RetriesRemaining),
	}
	if j.Variables != nil {
		raw, err := protojson.Marshal(j.Variables)
		if err == nil {
			sdkJob.Variables = json.RawMessage(raw)
		}
	}
	if j.LockExpiresAt != nil {
		t := j.LockExpiresAt.AsTime()
		sdkJob.LockExpiresAt = &t
	}
	return sdkJob
}

// marshalStruct converts a map to a structpb.Struct.
func marshalStruct(m map[string]any) (*structpb.Struct, error) {
	b, err := json.Marshal(m)
	if err != nil {
		return nil, err
	}
	var raw map[string]any
	if err := json.Unmarshal(b, &raw); err != nil {
		return nil, err
	}
	return structpb.NewStruct(raw)
}

File: workflow-sdk-python/src/workflow_sdk/client/grpc.py
"""gRPC transport implementation of EngineClient (worker-facing operations only)."""

from __future__ import annotations

from typing import Any

import grpc
from google.protobuf import json_format
from google.protobuf import struct_pb2

from workflow_sdk.internal.gen.workflow.v1 import engine_pb2
from workflow_sdk.internal.gen.workflow.v1 import engine_pb2_grpc


class GrpcEngineClient:
    """Communicates with the workflow engine over gRPC.

    Only implements the three operations used by the worker runner:
    poll_jobs, complete_job, and fail_job.

    Args:
        target: gRPC server address, e.g. ``"localhost:9090"``.
    """

    def __init__(self, target: str) -> None:
        self._channel = grpc.insecure_channel(target)
        self._stub = engine_pb2_grpc.WorkflowEngineStub(self._channel)

    def poll_jobs(
        self,
        worker_id: str,
        job_types: list[str],
        max_jobs: int = 1,
    ) -> list[dict[str, Any]]:
        req = engine_pb2.PollJobsRequest(
            worker_id=worker_id,
            job_types=job_types,
            max_jobs=max_jobs,
        )
        resp = self._stub.PollJobs(req)
        return [_proto_job_to_dict(j) for j in resp.jobs]

    def complete_job(
        self,
        job_id: str,
        worker_id: str,
        variables: dict[str, Any] | None = None,
    ) -> None:
        req = engine_pb2.CompleteJobRequest(job_id=job_id, worker_id=worker_id)
        if variables:
            s = struct_pb2.Struct()
            s.update(variables)
            req.variables_to_set.CopyFrom(s)
        self._stub.CompleteJob(req)

    def fail_job(
        self,
        job_id: str,
        worker_id: str,
        error_message: str,
        retryable: bool = True,
    ) -> None:
        self._stub.FailJob(engine_pb2.FailJobRequest(
            job_id=job_id,
            worker_id=worker_id,
            error_message=error_message,
            retryable=retryable,
        ))

    def close(self) -> None:
        self._channel.close()


def _proto_job_to_dict(job: Any) -> dict[str, Any]:
    variables: dict[str, Any] = {}
    if job.HasField("variables"):
        variables = json_format.MessageToDict(job.variables)
    return {
        "id": job.id,
        "jobType": job.job_type,
        "instanceId": job.instance_id,
        "stepExecutionId": job.step_execution_id,
        "retriesRemaining": job.retries_remaining,
        "variables": variables,
    }
